netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -813,116 +852,239 @@ _net_ebpf_extension_connection_context_initialize(
+
+Exit:
+    return result;
+}
+
 static ebpf_result_t
-_net_ebpf_ext_insert_connection_context_to_list(_Inout_ net_ebpf_extension_connection_context_t* connection_context)
+_net_ebpf_ext_insert_connection_context_to_list(
+    _In_ const bpf_sock_addr_t* sock_addr_ctx, UINT64 transport_endpoint_handle)
 {
     ebpf_result_t result = EBPF_SUCCESS;
-    KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+    bool locked = false;
+    KIRQL old_irql = PASSIVE_LEVEL;
+    net_ebpf_extension_connection_context_t* blocked_connection_context =
+        (net_ebpf_extension_connection_context_t*)ExAllocatePoolUninitialized(
+            NonPagedPoolNx, sizeof(net_ebpf_extension_connection_context_t), NET_EBPF_EXTENSION_POOL_TAG);
+
+    NET_EBPF_EXT_BAIL_ON_ALLOC_FAILURE_RESULT(
+        NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR, blocked_connection_context, "blocked_connection", result);
+    memset(blocked_connection_context, 0, sizeof(net_ebpf_extension_connection_context_t));
+
+    _net_ebpf_extension_connection_context_initialize(
+        sock_addr_ctx,
+        transport_endpoint_handle,
+        CONNECTION_CONTEXT_INITIALIZATION_SET_TIMESTAMP,
+        blocked_connection_context);
+
+    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+    locked = true;
 
     // Insert into hash table
     result = ebpf_hash_table_update(
-        _net_ebpf_ext_connect_context_hash_table,
-        (uint8_t*)connection_context,
-        (uint8_t*)&connection_context,
+        _net_ebpf_ext_blocked_connect_context_hash_table,
+        (uint8_t*)blocked_connection_context,
+        (uint8_t*)&blocked_connection_context,
         EBPF_HASH_TABLE_OPERATION_ANY);
     NET_EBPF_EXT_BAIL_ON_ERROR_RESULT(result);
 
-    // Insert the most recent entry at the head.
-    InsertHeadList(&_net_ebpf_ext_connect_context_list, &connection_context->list_entry);
-    _net_ebpf_ext_connect_context_count++;
-
-    // Purge stale entries from the list.
-    _net_ebpf_ext_purge_lru_contexts_under_lock(FALSE);
+    // Successfully inserted into the hash table. Also insert into the LRU list to ensure
+    // entries are not leaked.
+    InsertHeadList(&_net_ebpf_ext_blocked_connect_context_lru_list, &blocked_connection_context->list_entry);
+    ++_net_ebpf_ext_connect_context_count;
+    blocked_connection_context = NULL;
+    InterlockedIncrement(&_net_ebpf_ext_statistics.block_connection_count);
 
 Exit:
+    // If the lock was not acquired yet, acquire it now. This is necessary for using the low-memory list
+    // and for purging stale entries.
+    if (!locked) {
+        old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+        locked = true;
+    }
+    if (result != EBPF_SUCCESS) {
+        // If any failure occurred, attempt to use low memory list instead.
+        result = _net_ebpf_ext_insert_connection_context_to_low_memory_list(sock_addr_ctx, transport_endpoint_handle);
```

Comment:
nit: Should we increment / decrement `_net_ebpf_ext_connect_context_count` and `_net_ebpf_ext_statistics.block_connection_count` in this case too?

Also, do you think it makes sense to maintain a global counter which tracks how many times netebpfext ever had to fall back to preallocated contexts?

Reply:
Added two new counters - one in statistics which increases count whenever we use a pre allocated entry, and then a second counter for the in-use low-memory block connect_context_counts.